src/vs/workbench/api/common/extHostNotebook.ts

```diff
@@ -313,83 +313,91 @@ export class ExtHostNotebookController implements ExtHostNotebookShape {
-		if (token.isCancellationRequested) {
-			throw new CancellationError();
-		}
+			// validate write
+			await this._validateWriteFile(uri, options);
 
-		// Don't accept any cancellation beyond this point, we need to report the result of the file write
-		this.trace(`serialized versionId: ${versionId} ${uri.toString()}`);
-		await this._extHostFileSystem.value.writeFile(uri, bytes);
-		this.trace(`Finished write versionId: ${versionId} ${uri.toString()}`);
-		const providerExtUri = this._extHostFileSystem.getFileSystemProviderExtUri(uri.scheme);
-		const stat = await this._extHostFileSystem.value.stat(uri);
+			if (token.isCancellationRequested) {
+				throw new CancellationError();
+			}
+			const bytes = await serializer.serializer.serializeNotebook(data, token);
+			if (token.isCancellationRequested) {
+				throw new CancellationError();
+			}
 
-		const fileStats = {
-			name: providerExtUri.basename(uri),
-			isFile: (stat.type & files.FileType.File) !== 0,
-			isDirectory: (stat.type & files.FileType.Directory) !== 0,
-			isSymbolicLink: (stat.type & files.FileType.SymbolicLink) !== 0,
-			mtime: stat.mtime,
-			ctime: stat.ctime,
-			size: stat.size,
-			readonly: Boolean((stat.permissions ?? 0) & files.FilePermission.Readonly) || !this._extHostFileSystem.value.isWritableFileSystem(uri.scheme),
-			locked: Boolean((stat.permissions ?? 0) & files.FilePermission.Locked),
-			etag: files.etag({ mtime: stat.mtime, size: stat.size }),
-			children: undefined
-		};
+			// Don't accept any cancellation beyond this point, we need to report the result of the file write
+			this.trace(`serialized versionId: ${versionId} ${uri.toString()}`);
+			await this._extHostFileSystem.value.writeFile(uri, bytes);
+			this.trace(`Finished write versionId: ${versionId} ${uri.toString()}`);
+			const providerExtUri = this._extHostFileSystem.getFileSystemProviderExtUri(uri.scheme);
+			const stat = await this._extHostFileSystem.value.stat(uri);
+
+			const fileStats = {
+				name: providerExtUri.basename(uri),
+				isFile: (stat.type & files.FileType.File) !== 0,
+				isDirectory: (stat.type & files.FileType.Directory) !== 0,
+				isSymbolicLink: (stat.type & files.FileType.SymbolicLink) !== 0,
+				mtime: stat.mtime,
+				ctime: stat.ctime,
+				size: stat.size,
+				readonly: Boolean((stat.permissions ?? 0) & files.FilePermission.Readonly) || !this._extHostFileSystem.value.isWritableFileSystem(uri.scheme),
+				locked: Boolean((stat.permissions ?? 0) & files.FilePermission.Locked),
+				etag: files.etag({ mtime: stat.mtime, size: stat.size }),
+				children: undefined
+			};
 
-		this.trace(`exit saveNotebook(versionId: ${versionId}, ${uri.toString()})`);
-		return fileStats;
+			this.trace(`exit saveNotebook(versionId: ${versionId}, ${uri.toString()})`);
+			return fileStats;
+		} catch (error) {
+			// return fileOperationsErrors to keep the whole object across serialization, these errors are handled specially by the WCS
```

Comment:
Cool, but I wish our IPC would make this easier for you so you don't have to do all this manually